### PR TITLE
HGCAL trigger: towers update

### PR DESCRIPTION
**Pull request by jbsauvan**
* Implement a more flexible way of defining trigger towers from trigger cells, based on arbitrary mappings.
* Small fix in GenTau ntuplizer

Depends on external cms-data/L1Trigger-L1THGCal#8

### CODE REVIEW

File: DataFormats/L1THGCal/interface/HGCalTowerID.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+    const unsigned short rawId;
+    const float eta;
+    const float phi;
+
+  };
+
+
+}
+
```
Comment (kpedro88) [inline DataFormats/L1THGCal/interface/HGCalTowerID.h]: remove unneeded whitespace

File: DataFormats/L1THGCal/interface/HGCalTowerMap.h [hunk 2]
```diff
@@ -7,60 +7,41 @@
...
 
-    HGCalTowerMap(): nEtaBins_(0), nPhiBins_(0), layer_(0) {}
-    
-    HGCalTowerMap( int nEtaBins, int nPhiBins );
+    HGCalTowerMap(): layer_(0) {}
+
+    HGCalTowerMap(const std::vector<l1t::HGCalTowerCoord>& tower_ids, const int layer);
 
-    HGCalTowerMap( const std::vector<double>& etaBins, const std::vector<double>& phiBins );
+    // ~HGCalTowerMap();
```
Comment (kpedro88) [inline DataFormats/L1THGCal/interface/HGCalTowerMap.h]: delete commented-out code

File: DataFormats/L1THGCal/interface/HGCalTowerMap.h [hunk 3]
```diff
@@ -7,60 +7,41 @@
...
 
-    ~HGCalTowerMap();
+    // const l1t::HGCalTower& tower(int iX, int iY) const;
```
Comment (kpedro88) [inline DataFormats/L1THGCal/interface/HGCalTowerMap.h]: delete commented-out code

File: DataFormats/L1THGCal/interface/HGCalTowerMap.h [hunk 4]
```diff
@@ -7,60 +7,41 @@
...
+    int layer() const { return layer_; }
 
+    const HGCalTowerMap& operator+=(const HGCalTowerMap& map);
 
-    int nEtaBins() const { return nEtaBins_; }
-    int nPhiBins() const { return nPhiBins_; }
-    const vector<double>& etaBins() const { return etaBins_; }
-    const vector<double>& phiBins() const { return phiBins_; }
-    const l1t::HGCalTower& tower(int iEta, int iPhi) const { return towerMap_.at(bin_id(iEta,iPhi)); }
+    // bool addEt(short iX, short iY, float etEm, float etHad);
```
Comment (kpedro88) [inline DataFormats/L1THGCal/interface/HGCalTowerMap.h]: delete commented-out code

File: DataFormats/L1THGCal/src/HGCalTower.cc [hunk 5]
```diff
@@ -1,100 +1,57 @@
...
   }
-
-  this->setP4(this->p4() + tower.p4());
-  this->setEtEm(this->etEm() + tower.etEm());
-  this->setEtHad(this->etHad() + tower.etHad());
-
-  this->setHwPt(this->hwPt() + tower.hwPt());
-  this->setHwEtEm(this->hwEtEm() + tower.hwEtEm());
-  this->setHwEtHad(this->hwEtHad() + tower.hwEtHad());
+  this->setP4(PolarLorentzVector(this->pt()+tower.pt(), this->eta(), this->phi(), 0.));
```
Comment (kpedro88) [inline DataFormats/L1THGCal/src/HGCalTower.cc]: you could use the `addEt()` function here

File: DataFormats/L1THGCal/src/HGCalTowerMap.cc [hunk 6]
```diff
@@ -1,153 +1,45 @@
...
-    if(phi_norm>=phiBins_[i] && phi_norm<phiBins_[i+1]){
-      return i;
-    }
-  }
-  
-  return -1;
-
 }
 
+// HGCalTowerMap::~HGCalTowerMap() {}
```
Comment (kpedro88) [inline DataFormats/L1THGCal/src/HGCalTowerMap.cc]: delete commented-out code

File: DataFormats/L1THGCal/src/HGCalTowerMap.cc [hunk 7]
```diff
@@ -1,153 +1,45 @@
...
 
-  for(int iEta=-nEtaBins_; iEta<=nEtaBins_; iEta++){
-    if(iEta==0) continue;
-    for(int iPhi=0; iPhi<nPhiBins_; iPhi++){
-      towerMap_[bin_id(iEta,iPhi)] += map.tower(iEta,iPhi);
+  for(auto tower: map.towers()) {
+    auto this_tower = towerMap_.find(tower.first);
+    if(this_tower == towerMap_.end()) {
+      throw edm::Exception(edm::errors::StdException, "StdException")
+        << "HGCalTowerMap: Trying to add HGCalTowerMaps but cound not find bin: " << tower.first <<endl;
```
Comment (kpedro88) [inline DataFormats/L1THGCal/src/HGCalTowerMap.cc]: typo cound -> could

File: L1Trigger/L1THGCal/python/hgcalTriggerGeometryESProducer_cfi.py [hunk 8]
```diff
@@ -18,6 +18,7 @@
         ]
 
 
+
```
Comment (kpedro88) [inline L1Trigger/L1THGCal/python/hgcalTriggerGeometryESProducer_cfi.py]: revert unnecessary whitespace-only changes to this file

File: L1Trigger/L1THGCal/src/HGCalTriggerTowerGeometryHelper.cc [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+                                                                                                  maxPhi_(conf.getParameter<double>("maxPhi")),
+                                                                                                  nBinsEta_(conf.getParameter<int>("nBinsEta")),
+                                                                                                  nBinsPhi_(conf.getParameter<int>("nBinsPhi")),
+                                                                                                  binsEta_(conf.getParameter<std::vector<double> >("binsEta")),
+                                                                                                  binsPhi_(conf.getParameter<std::vector<double> >("binsPhi")) {
+
+
+  if(!binsEta_.empty() && ((unsigned int)(binsEta_.size()) != nBinsEta_+1)) {
+    throw edm::Exception(edm::errors::Configuration, "Configuration")
+      << "HGCalTriggerTowerGeometryHelper nBinsEta for the tower map not consistent with binsEta size"<<std::endl;
```
Comment (kpedro88) [inline L1Trigger/L1THGCal/src/HGCalTriggerTowerGeometryHelper.cc]: why not just replace the value of `nBinsEta_` with `binsEta_.size()` if `binsEta_` is specified?

File: L1Trigger/L1THGCal/src/HGCalTriggerTowerGeometryHelper.cc [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+    for(; l1tTriggerTowerMappingStream >> trigger_cell_id >> iEta >> iPhi;) {
+      if(iEta >= nBinsEta_ || iPhi >= nBinsPhi_) {
+        throw edm::Exception(edm::errors::Configuration, "Configuration")
+          << "HGCalTriggerTowerGeometryHelper warning inconsistent mapping TC : " << trigger_cell_id
+          << " to TT iEta: " << iEta << " iPhi: " << iPhi
+          << " when max #bins eta: "  << nBinsEta_ << " phi: " << nBinsPhi_ << std::endl;
+      }
+      HGCalDetId detId(trigger_cell_id);
+      int zside = detId.zside();
+      l1t::HGCalTowerID towerId(zside, iEta, iPhi);
```
Comment (kpedro88) [inline L1Trigger/L1THGCal/src/HGCalTriggerTowerGeometryHelper.cc]: could avoid the temporary with `l1t::HGCalTowerID towerId(HGCalDetId(trigger_cell_id).zside(), iEta, iPhi);`

File: L1Trigger/L1THGCal/src/HGCalTriggerTowerGeometryHelper.cc [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+    } else if(phi >= maxPhi_) {
+      bin_phi = 0;
+    } else {
+      edm::LogError("HGCalTriggerTowerGeometryHelper") << " did not manage to map TC " << trigger_cell_id << " (phi: " << phi << ") to any Trigger Tower\n";
+    }
+  } else {
+    bin_phi = bin_phi_l - binsPhi_.begin();
+  }
+  int zside = eta < 0 ?  -1 : 1;
+  // std::cout << "TC " << trigger_cell_id << " eta: " << eta << " phi: " << phi << " mapped to bin: (" << zside << ", " << bin_eta << ", " << bin_phi << ")" << std::endl;
```
Comment (kpedro88) [inline L1Trigger/L1THGCal/src/HGCalTriggerTowerGeometryHelper.cc]: delete commented-out code

File: L1Trigger/L1THGCal/src/HGCalTriggerTowerGeometryHelper.cc [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+  auto bin_eta_l = std::lower_bound(binsEta_.begin(), binsEta_.end(), fabs(eta));
+  unsigned int bin_eta = 0;
+  // we add a protection for TCs in Hadron part which are outside the boundaries and possible rounding effects
+  if(bin_eta_l == binsEta_.end()) {
+    if(fabs(eta) < minEta_) {
+      bin_eta = 0;
+    } else if(fabs(eta) >= maxEta_) {
+      bin_eta = nBinsEta_;
+    } else {
+      edm::LogError("HGCalTriggerTowerGeometryHelper") << " did not manage to map TC " << trigger_cell_id << " (eta: " << eta << ") to any Trigger Tower\n";
```
Comment (kpedro88) [inline L1Trigger/L1THGCal/src/HGCalTriggerTowerGeometryHelper.cc]: should it throw instead of just printing an error?

File: L1Trigger/L1THGCal/src/HGCalTriggerTowerGeometryHelper.cc (hunk 10 above)
Comment (kpedro88) [reply L1Trigger/L1THGCal/src/HGCalTriggerTowerGeometryHelper.cc]: however `HGCalDetId` won't work for the new geometry

File: L1Trigger/L1THGCal/test/HGCalTriggerGeomTester.cc [hunk 13]
```diff
@@ -26,10 +26,10 @@
 
 #include "L1Trigger/L1THGCal/interface/HGCalTriggerGeometryBase.h"
 
-#include <stdlib.h> 
+#include <stdlib.h>
```
Comment (kpedro88) [inline L1Trigger/L1THGCal/test/HGCalTriggerGeomTester.cc]: use `cstdlib`

**Comment by kpedro88** [comment]
Have you tested this privately with #23545?

**Comment by jbsauvan** [comment]
Hi @kpedro88 
I am currently testing the HGCAL trigger code with #23545 and checking what needs to be modified. The most critical part will be the cell -> trigger cell mapping. This trigger cell -> tower mapping should be then easier to adapt.

**Comment by jbsauvan** [comment]
@cerminar could you have a look at the comments?
Thanks!

**Comment by mrodozov** [comment]
please test with cms-sw/cmsdist#4145

File: L1Trigger/L1THGCal/src/HGCalTriggerTowerGeometryHelper.cc (hunk 9 above)
Comment (cerminar) [reply L1Trigger/L1THGCal/src/HGCalTriggerTowerGeometryHelper.cc]: @kpedro88: I don't particularly like overriding user parameters. I think it is better to enforce consistency. Note however that this was already the behavior of the previous version of the code.

File: L1Trigger/L1THGCal/src/HGCalTriggerTowerGeometryHelper.cc (hunk 12 above)
Comment (cerminar) [reply L1Trigger/L1THGCal/src/HGCalTriggerTowerGeometryHelper.cc]: @kpedro88 I think that the error is ok for now. Maybe we will review this decision once we have some more experience with this code.

**Comment by kpedro88** [comment]
@jbsauvan @cerminar several comments remain unaddressed:
https://github.com/cms-sw/cmssw/pull/23608#discussion_r196100354
https://github.com/cms-sw/cmssw/pull/23608#discussion_r196100639
https://github.com/cms-sw/cmssw/pull/23608#discussion_r196101636

**Comment by kpedro88** [comment]
please test with cms-sw/cmsdist#4145

**Comment by jbsauvan** [comment]
Is there something else needed from our side?

**Comment by kpedro88** [comment]
please test with cms-sw/cmsdist#4145

**Comment by kpedro88** [comment]
+1

**Comment by rekovic** [comment]
+1

**Comment by kpedro88** [comment]
@fabiocos can this go in?

**Comment by fabiocos** [comment]
@kpedro88 in which test workflow is this code really running?

**Comment by kpedro88** [comment]
@fabiocos as far as I know, it runs in all Phase2 workflows...

**Comment by kpedro88** [comment]
(except the new HGCal v9 workflows)

**Comment by fabiocos** [comment]
+1

affects only Phase2 workflows
